### PR TITLE
LUCENE-10262: Lift up restrictions for navigating PointValues#PointTree

### PR DESCRIPTION
**Pull request by iverase**
see https://issues.apache.org/jira/browse/LUCENE-10262

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/util/bkd/BKDReader.java [hunk 1]
```diff
@@ -515,15 +529,20 @@ public void addAll(PointValues.IntersectVisitor visitor, boolean grown) throws I
 
     @Override
     public void visitDocValues(PointValues.IntersectVisitor visitor) throws IOException {
+      maybeResetNodeDataPosition();
+      addOneByOne(visitor);
+    }
+
+    public void addOneByOne(PointValues.IntersectVisitor visitor) throws IOException {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/util/bkd/BKDReader.java:536]: Maybe give it a more explicit name, e.g. `visitLeavesOneByOne`?

**Review by jpountz — APPROVED** [review]
It makes the API much easier to use!

File: lucene/core/src/java/org/apache/lucene/util/bkd/BKDReader.java [hunk 2]
```diff
@@ -297,11 +301,20 @@ public boolean moveToChild() throws IOException {
...
       pushLeft();
       return true;
     }
 
+    private void maybeResetNodeDataPosition() throws IOException {
+      if (readNodeDataPositions[level] != innerNodes.getFilePointer()) {
+        // move position of the inner nodes index in case we move to the first child
+        assert readNodeDataPositions[level] < innerNodes.getFilePointer();
+        innerNodes.seek(readNodeDataPositions[level]);
+      }
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/util/bkd/BKDReader.java]: yes, quick check and it makes no difference.

File: lucene/core/src/java/org/apache/lucene/util/bkd/BKDReader.java (hunk 1 above)
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/util/bkd/BKDReader.java]: done

**Review by jpountz — APPROVED** [review]

**Pull request merged**